src/compartment_sets.cpp

```diff
@@ -0,0 +1,669 @@
+
+
+#include "../extlib/filesystem.hpp"
+
+#include "utils.h"  // readFile
+
+#include <unordered_set>
+
+#include <bbp/sonata/compartment_sets.h>
+namespace bbp {
+namespace sonata {
+
+namespace fs = ghc::filesystem;
+
+namespace detail {
+
+using json = nlohmann::json;
+
+
+class CompartmentLocation
+{
+    private:
+        std::uint64_t gid_;
+        std::uint64_t section_idx_;
+        double offset_;
+
+        void setGid(int64_t gid) {
+            if (gid < 0) {
+                throw SonataError(fmt::format("GID must be non-negative, got {}", gid));
+            }
+            gid_ = static_cast<uint64_t>(gid);
+        }
+        void setSectionIdx(int64_t section_idx) {
+            if (section_idx < 0) {
+                throw SonataError(
+                    fmt::format("Section index must be non-negative, got {}", section_idx));
+            }
+            section_idx_ = static_cast<uint64_t>(section_idx);
+        }
+        void setOffset(double offset) {
+            if (offset < 0.0 || offset > 1.0) {
+                throw SonataError(
+                    fmt::format("Offset must be between 0 and 1 inclusive, got {}", offset));
+            }
+            offset_ = offset;
+        }
+
+  public:
+    static constexpr double offsetTolerance = 1e-4;
+    // static constexpr double offsetToleranceInv = 1.0 / offsetTolerance;
+
+    CompartmentLocation(const CompartmentLocation& other) = default;
+    CompartmentLocation(CompartmentLocation&&) noexcept = default;
+    CompartmentLocation& operator=(CompartmentLocation&&) noexcept = default;
+    CompartmentLocation(int64_t gid, int64_t section_idx, double offset) {
+        setGid(gid);
+        setSectionIdx(section_idx);
+        setOffset(offset);
+    }
+
+    CompartmentLocation(const std::string& content)
+        : CompartmentLocation(json::parse(content)) {}
+
+    CompartmentLocation(const nlohmann::json& j) {
+        if (!j.is_array() || j.size() != 3) {
+            throw SonataError(
+                "CompartmentLocation must be an array of exactly 3 elements: [gid, section_idx, "
+                "offset]");
+        }
+
+        setGid(get_int64_or_throw(j[0]));
+        setSectionIdx(get_int64_or_throw(j[1]));
+
+        if (!j[2].is_number()) {
+            throw SonataError("Fourth element (offset) must be a number");
+        }
+        setOffset(j[2].get<double>());
+    }
+
+    uint64_t gid() const {
+        return gid_;
+    }
+
+    uint64_t sectionIdx() const {
+        return section_idx_;
+    }
+
+    double offset() const {
+        return offset_;
+    }
+
+    nlohmann::json to_json() const {
+        return nlohmann::json::array({gid_, section_idx_, offset_});
+    }
+
+    bool operator==(const CompartmentLocation& other) const {
+        return gid_ == other.gid_ && section_idx_ == other.section_idx_ &&
+               std::abs(offset_ - other.offset_) < offsetTolerance;
```

Comment:
In this case, I think the floats would be exactly equal, so no tolerance testing is required.

I still wonder under what circumstances one will be comparing these values, though.

Reply:
I still don't think you want a tolerance, especially one of `1e-4` - other than Locations that are *exactly* the same, I think it's best to make the recording exactly as the user asked for.